skbio/metadata/_interval.py

```diff
@@ -0,0 +1,599 @@
+
+            if self.dropped:
+                raise RuntimeError('Cannot change boundaries to dropped '
+                                   'Interval object.')
+            self._boundaries = value
+
+    @property
+    @experimental(as_of='0.4.2-dev')
+    def locations(self):
+        '''The coordinates of the interval feature.
+
+        It should be a list of tuples of int pair. Each tuple stores
+        the start and end coordinates of a span of the interval feature.
+        '''
+        if self.dropped:
+            raise RuntimeError(
+                'Cannot retrieve locations from dropped Interval object.')
+        return self._locations
+
+    @locations.setter
+    @experimental(as_of='0.4.2-dev')
+    def locations(self, value):
+        if value is None:
+            self._locations = None
+        else:
+            for location in value:
+                _assert_valid_location(location)
+
+            if self.dropped:
+                raise RuntimeError(
+                    'Cannot change locations to dropped Interval object.')
+            self._locations = value
+            if self._interval_metadata is not None:
+                self._interval_metadata._is_stale_tree = True
+
+    @property
+    @experimental(as_of='0.4.2-dev')
+    def metadata(self):
+        '''The metadata of the interval feature.
+
+        It stores the metadata (eg. gene name, function, ID, etc.) of
+        the interval feature as a dict.
+        '''
+        if self.dropped:
+            raise RuntimeError(
+                'Cannot retrieve metadata from dropped Interval object.')
+        return self._metadata
+
+    @metadata.setter
+    @experimental(as_of='0.4.2-dev')
+    def metadata(self, value):
+        if self.dropped:
+            raise RuntimeError('Cannot change metadata to dropped '
+                               'Interval object.')
+        if not isinstance(value, dict):
+            raise TypeError("metadata must be a dict, not %r" % value)
+        self._metadata = value
+
+    @property
+    @experimental(as_of='0.4.2-dev')
+    def dropped(self):
+        '''Boolean value indicating if the ``Interval`` object is dropped.
+
+        Notes
+        -----
+        This property is not writable.
+
+        See Also
+        --------
+        skbio.metadata.Interval.drop
+        '''
+        return self._interval_metadata is None
+
+
+class IntervalMetadata():
+    """Stores the interval features of a sequence as a list of ``Interval`` objects.
+
+    ``IntervalMetadata`` object to allow for the storage, modification, and
+    querying of interval features covering a region of a single coordinate
+    system. For instance, this can be used to store functional annotations
+    about genes across a genome.
+
+    This object is typically coupled with another object, such as a ``Sequence``
+    object, or a ``TabularMSA`` object.
+
+    See Also
+    --------
+    skbio.metadata.Interval
+
+    Examples
+    --------
```

Comment:
the reason I didn't break it into multiple section is to avoid repeating the code of setting up `im`. I can break it.